front_end/models/react_native/ReactDevToolsBindingsModel.ts

```diff
@@ -115,23 +113,118 @@ export class ReactDevToolsBindingsModel extends SDK.SDKModel.SDKModel {
+        const possiblyError = response.getError();
+        if (possiblyError) {
+          throw new Error('Failed to add binding for ReactDevToolsBindingsModel: ' + possiblyError);
+        }
+
+        this.enabled = true;
+        this.initializeExecutionContextListeners();
+      });
+  }
+
+  isEnabled(): boolean {
+    return this.enabled;
+  }
+
+  private initializeExecutionContextListeners(): void {
+    const runtimeModel = this.target().model(SDK.RuntimeModel.RuntimeModel);
+    if (!runtimeModel) {
+      throw new Error('Failed to initialize execution context listeners for ReactDevToolsBindingsModel: runtime model is not available');
+    }
+
+    runtimeModel.addEventListener(SDK.RuntimeModel.Events.ExecutionContextCreated, this.onExecutionContextCreated, this);
+    runtimeModel.addEventListener(SDK.RuntimeModel.Events.ExecutionContextDestroyed, this.onExecutionContextDestroyed, this);
+  }
+
+  private onExecutionContextCreated({data: executionContext}: ContextCreatedEvent): void {
+    if (executionContext.name !== MAIN_EXECUTION_CONTEXT_NAME) {
+      return;
+    }
+
+    void this.waitForFuseboxDispatcherToBeInitialized()
+      .then(() => this.dispatchEventToListeners(Events.BackendExecutionContextCreated))
+      .catch((error: Error) => this.dispatchEventToListeners(Events.BackendExecutionContextUnavailable, error.message));
+  }
+
+  private onExecutionContextDestroyed({data: executionContext}: ContextDestroyedEvent): void {
+    if (executionContext.name !== MAIN_EXECUTION_CONTEXT_NAME) {
+      return;
+    }
+
+    this.dispatchEventToListeners(Events.BackendExecutionContextDestroyed);
+  }
+
+  private async waitForFuseboxDispatcherToBeInitialized(attempt = 1): Promise<void> {
+    // Ideally, this should not be polling, but rather one `Runtime.evaluate` request with `awaitPromise` option
+    // We need to support it in Hermes first, then we can migrate this to awaitPromise
+    if (attempt >= 20) { // ~5 seconds
+      throw new Error('Failed to wait for Fusebox Dispatcher initialization: it took too long');
+    }
+
+    const runtimeModel = this.target().model(SDK.RuntimeModel.RuntimeModel);
     if (!runtimeModel) {
-      throw new Error('[ReactDevToolsBindingsModel] Failed to enable model: runtime model is not available');
+      throw new Error('Failed to wait for Fusebox Dispatcher initialization: runtime model is not available');
     }
 
-    await runtimeModel.agent.invoke_evaluate({expression: `${RUNTIME_GLOBAL}.BINDING_NAME`})
-      .then(response => response.result.value)
-      .then(bindingName => runtimeModel.agent.invoke_addBinding({name: bindingName}));
+    await runtimeModel.agent.invoke_evaluate({
+      expression: `globalThis.${RUNTIME_GLOBAL} != undefined`,
+      returnByValue: true,
+    })
+      .then(response => {
+        if (response.exceptionDetails) {
+          throw new Error('Failed to wait for Fusebox Dispatcher initialization: ' + response.exceptionDetails.text);
```

Comment:
Can we genericise/tweak this message? I'd rather we hide the Fusebox codename from users.

```suggestion
          throw new Error('Failed to wait for dispatcher initialization (ReactDevToolsBindingsModel): ' + response.exceptionDetails.text);
```

Reply:
Done!